File: MetaGeta/MetaGeta.DataStore/JobQueue/JobQueue.vb
Public Class JobQueue
    Implements IDisposable, INotifyPropertyChanged

    Private Shared ReadOnly log As log4net.ILog = log4net.LogManager.GetLogger(Reflection.MethodBase.GetCurrentMethod().DeclaringType)

    Private ReadOnly m_ActionWaitingQueue As New Queue(Of Job)
    Private ReadOnly m_AllActionItems As New List(Of Job)

    Private ReadOnly m_ProcessActionThread As Thread
    Private ReadOnly m_StopProcessingActions As New ManualResetEvent(False)
    Private ReadOnly m_QueueThreadIsSleeping As New ManualResetEvent(False)

    Public Sub New()
        m_ProcessActionThread = New Thread(AddressOf ProcessActionQueueThread)
        m_ProcessActionThread.Start()
    End Sub

    Public ReadOnly Property ActionItems() As IEnumerable(Of Job)
        Get
            SyncLock m_ActionWaitingQueue
                Return m_AllActionItems.ToArray()
            End SyncLock
        End Get
    End Property

    Public Sub EnqueueAction(ByVal action As String, ByVal dataStore As MGDataStore, ByVal file As MGFile)
        SyncLock m_ActionWaitingQueue
            Dim item As New Job(action, dataStore, file)
            m_ActionWaitingQueue.Enqueue(item)
            Monitor.PulseAll(m_ActionWaitingQueue)
            m_AllActionItems.Add(item)
        End SyncLock
        OnActionItemsChanged()
    End Sub

    Private Sub CloseActionQueue()
        log.Info("Stopping action queue...")
        SyncLock m_ActionWaitingQueue
            m_ActionWaitingQueue.Enqueue(Nothing)
            Monitor.Pulse(m_ActionWaitingQueue)
        End SyncLock
        m_StopProcessingActions.Set()
        log.Info("Joining...")
        m_ProcessActionThread.Join()
        log.Info("Joined")
    End Sub

    Private Sub ProcessActionQueueThread()
        Thread.CurrentThread.Name = "ProcessActionQueue"
        While True
            If m_StopProcessingActions.WaitOne(0, False) Then Return

            Dim nextItem As Job
            SyncLock m_ActionWaitingQueue
                While m_ActionWaitingQueue.Count = 0
                    m_QueueThreadIsSleeping.Set()
                    Monitor.Wait(m_ActionWaitingQueue)
                    m_QueueThreadIsSleeping.Reset()
                End While
                nextItem = m_ActionWaitingQueue.Dequeue()
            End SyncLock
            If nextItem Is Nothing Then Return

            Dim action = nextItem.DataStore.LookupAction(nextItem.Action)
            nextItem.Status.Start()
            Try
                log.DebugFormat("Starting action ""{0}"" for file ""{1}""...", nextItem.Action, If(nextItem.File IsNot Nothing, nextItem.File.FileName, ""))
                action.DoAction(nextItem.Action, nextItem.File, nextItem.Status)
                nextItem.Status.Done(True)
                nextItem.Status.StatusMessage = String.Empty
                log.DebugFormat("Action ""{0}"" for file ""{1}"" done.", nextItem.Action, If(nextItem.File IsNot Nothing, nextItem.File.FileName, ""))
            Catch ex As Exception
                nextItem.Status.Done(False)
                nextItem.Status.StatusMessage = ex.ToString()
                log.ErrorFormat("Action ""{0}"" for file ""{1}"" failed with exception:\n{2}", nextItem.Action, If(nextItem.File IsNot Nothing, nextItem.File.FileName, ""), ex)
            End Try
        End While
    End Sub

    Public Sub WaitForQueueToEmpty()
        While True
            SyncLock m_ActionWaitingQueue
                'Prevents race condition were worker hasn't started yet
                If m_ActionWaitingQueue.Count = 0 Then Exit While
            End SyncLock
            Thread.Sleep(100)
        End While
        m_QueueThreadIsSleeping.WaitOne()
    End Sub

    Private Sub OnActionItemsChanged()
        OnPropertyChanged("ActionItems")
    End Sub

    Private Sub OnPropertyChanged(ByVal propertyName As String)
        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
    End Sub

    Public Event PropertyChanged(ByVal sender As Object, ByVal e As System.ComponentModel.PropertyChangedEventArgs) Implements System.ComponentModel.INotifyPropertyChanged.PropertyChanged


#Region " IDisposable Support "
    Private disposedValue As Boolean = False        ' To detect redundant calls

    ' IDisposable
    Protected Overridable Sub Dispose(ByVal disposing As Boolean)
        If Not Me.disposedValue Then
            If disposing Then
                CloseActionQueue()
            End If

            ' TODO: free your own state (unmanaged objects).
            ' TODO: set large fields to null.
        End If
        Me.disposedValue = True
    End Sub

    ' This code added by Visual Basic to correctly implement the disposable pattern.
    Public Sub Dispose() Implements IDisposable.Dispose
        ' Do not change this code.  Put cleanup code in Dispose(ByVal disposing As Boolean) above.
        Dispose(True)
        GC.SuppressFinalize(Me)
    End Sub
#End Region

End Class

File: MetaGeta/MetaGeta.DataStore/JobQueue/JobQueue.cs
// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Reflection;
using System.Threading;
using log4net;

#endregion

namespace MetaGeta.DataStore {
    public class JobQueue : IDisposable, INotifyPropertyChanged {
        private static readonly ILog log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
        private readonly Queue<Job> m_ActionWaitingQueue = new Queue<Job>();

        private readonly List<Job> m_AllActionItems = new List<Job>();
        private readonly Thread m_ProcessActionThread;

        private readonly ManualResetEvent m_QueueThreadIsSleeping = new ManualResetEvent(false);
        private readonly ManualResetEvent m_StopProcessingActions = new ManualResetEvent(false);

        public JobQueue(bool designMode) {
            if (!designMode) {
                m_ProcessActionThread = new Thread(ProcessActionQueueThread);
                m_ProcessActionThread.Start();
            }
        }

        public IEnumerable<Job> ActionItems {
            get {
                lock (m_ActionWaitingQueue) {
                    return m_AllActionItems.ToArray();
                }
            }
        }

        #region INotifyPropertyChanged Members

        public event PropertyChangedEventHandler PropertyChanged;

        #endregion

        #region " IDisposable Support "

        // To detect redundant calls
        private bool disposedValue = false;

        // IDisposable

        // This code added by Visual Basic to correctly implement the disposable pattern.
        public void Dispose() {
            // Do not change this code.  Put cleanup code in Dispose(ByVal disposing As Boolean) above.
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing) {
            if (!disposedValue) {
                if (disposing)
                    CloseActionQueue();

                // TODO: free your own state (unmanaged objects).
                // TODO: set large fields to null.
            }
            disposedValue = true;
        }

        #endregion

        public void EnqueueAction(IAction action, MGDataStore dataStore, MGFile file) {
            lock (m_ActionWaitingQueue) {
                var item = new Job(action, dataStore, file);
                m_ActionWaitingQueue.Enqueue(item);
                Monitor.PulseAll(m_ActionWaitingQueue);
                m_AllActionItems.Add(item);
            }
            OnActionItemsChanged();
        }

        private void CloseActionQueue() {
            log.Info("Stopping action queue...");
            lock (m_ActionWaitingQueue) {
                m_ActionWaitingQueue.Enqueue(null);
                Monitor.Pulse(m_ActionWaitingQueue);
            }
            m_StopProcessingActions.Set();
            log.Info("Joining...");
            m_ProcessActionThread.Join();
            log.Info("Joined");
        }

        private void ProcessActionQueueThread() {
            Thread.CurrentThread.Name = "ProcessActionQueue";
            while (true) {
                if (m_StopProcessingActions.WaitOne(0, false))
                    return;

                Job nextItem = null;
                lock (m_ActionWaitingQueue) {
                    while (m_ActionWaitingQueue.Count == 0) {
                        m_QueueThreadIsSleeping.Set();
                        Monitor.Wait(m_ActionWaitingQueue);
                        m_QueueThreadIsSleeping.Reset();
                    }
                    nextItem = m_ActionWaitingQueue.Dequeue();
                }
                if (nextItem == null)
                    return;

                nextItem.Status.Start();
                try {
                    log.DebugFormat("Starting action \"{0}\" for file \"{1}\"...", nextItem.Action, nextItem.File != null ? nextItem.File.FileName : "");
                    nextItem.Action.Execute(nextItem.File, nextItem.Status);
                    nextItem.Status.Done(true);
                    nextItem.Status.StatusMessage = string.Empty;
                    log.DebugFormat("Action \"{0}\" for file \"{1}\" done.", nextItem.Action, nextItem.File != null ? nextItem.File.FileName : "");
                } catch (Exception ex) {
                    nextItem.Status.Done(false);
                    nextItem.Status.StatusMessage = ex.ToString();
                    log.ErrorFormat("Action \"{0}\" for file \"{1}\" failed with exception:\\n{2}", nextItem.Action, nextItem.File != null ? nextItem.File.FileName : "", ex);
                }
            }
        }

        public void WaitForQueueToEmpty() {
            while (true) {
                lock (m_ActionWaitingQueue) {
                    //Prevents race condition were worker hasn't started yet
                    if (m_ActionWaitingQueue.Count == 0)
                        break; // TODO: might not be correct. Was : Exit While
                }
                Thread.Sleep(100);
            }
            m_QueueThreadIsSleeping.WaitOne();
        }

        private void OnActionItemsChanged() {
            OnPropertyChanged("ActionItems");
        }

        private void OnPropertyChanged(string propertyName) {
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}